### ISSUE TITLE
AnimatedSprite2d: SpriteFrames Editor

### ISSUE THREAD

**Issue by almusx** [godotengine/godot#76729]
### Godot version

v4.1.dev1.official [db1302637]

### System information

Windows 10, Compatibility Rendering Driver, GTX 1660 Ti (531.79)

### Issue description

The mouse cursor disappears when i'm clicking and holding the mouse button on the little arrows in the Frame Duration option and moving the mouse over the Edit Box (with the x1 text).

![image](https://user-images.githubusercontent.com/7002131/236234058-e90c859a-4676-47b4-b2ee-e1ca79a0b2bc.png)

### Steps to reproduce

1. Add AnimatedSprite2d node
2. Load spritesheet
3. Open SpriteFrames editor
4. Select a frame
5. click and hold the mouse button over the little up or down arrow on the Frame Duration Option
6. While holding the mouse button move left over the edit box.

### Minimal reproduction project

[FrameDurationTest.zip](https://github.com/godotengine/godot/files/11398451/FrameDurationTest.zip)

**Comment by AThousandShips** [comment]
Does this happen in other cases where that type of input box is used? The one with the little arrows?

**Comment by AThousandShips** [comment]
This appears to be an error in the windows code, will take a deeper look

**Comment by AThousandShips** [comment]
I'm not sure but this might be intentional, but might also be an error, it might incorrectly use `MOUSE_MODE_CAPTURED` instead of `CONFINED`

**Comment by almusx** [comment]
> Does this happen in other cases where that type of input box is used? The one with the little arrows? Like the FPS one to the left

changing the fps works as expected, clicking the little arrows works and holding the mouse button i can change the fps moving mouse up /down or left / right. The mouse cursor disapear while holding and became visible after releasing.

I tried again the frame duration arrows and holding the mouse button, the mouse cursor disapear while moving the mouse left/right, dont need to move over the edit box and the mouse is locked on the center of the screen, if i right click its shows to add node

**Comment by AThousandShips** [comment]
Al in what way is it working differently?

**Comment by almusx** [comment]
> Al in what way is it working differently?

after releasing tyhe mouse button, the mouse still not visible and locked on the center of the screen, i cant do anything, only alt+f4 to exit.

**Comment by almusx** [comment]
i tested again, if i Alt+tab or press the Windows Key, the mouse become visible but after i click on the godot it disapear again and get locked to the screen center, also holding the mouse button while clicking the arrows doesnt change the frame duration.

**Comment by dalexeev** [comment]
Duplicate of #75652 (was closed).

**Comment by AThousandShips** [comment]
It does capture the mouse which makes it invisible and centered, sounds like it might be failing to restore it properly

**Comment by AThousandShips** [comment]
For reference, this is where it is captured:
https://github.com/godotengine/godot/blob/8c729f0f34a92b92a1a8204e1d56b6ad05cfd27b/editor/gui/editor_spin_slider.cpp#L109-L112

And here are the poitns where it is released:
https://github.com/godotengine/godot/blob/8c729f0f34a92b92a1a8204e1d56b6ad05cfd27b/editor/gui/editor_spin_slider.cpp#L80-L85

https://github.com/godotengine/godot/blob/8c729f0f34a92b92a1a8204e1d56b6ad05cfd27b/editor/gui/editor_spin_slider.cpp#L468-L474

**Comment by kleonc** [comment]
> For reference, this is where it is captured:

@AThousandShips Not really, that's a `SpinBox`:
https://github.com/godotengine/godot/blob/8c729f0f34a92b92a1a8204e1d56b6ad05cfd27b/editor/plugins/sprite_frames_editor_plugin.cpp#L1803-L1811
https://github.com/godotengine/godot/blob/8c729f0f34a92b92a1a8204e1d56b6ad05cfd27b/editor/plugins/sprite_frames_editor_plugin.cpp#L1889

Seems like after capturing the mouse on drag start:
https://github.com/godotengine/godot/blob/8c729f0f34a92b92a1a8204e1d56b6ad05cfd27b/scene/gui/spin_box.cpp#L178-L189
it ends up not being released, for whatever reason this is not executed in this case:
https://github.com/godotengine/godot/blob/8c729f0f34a92b92a1a8204e1d56b6ad05cfd27b/scene/gui/spin_box.cpp#L168-L174
https://github.com/godotengine/godot/blob/8c729f0f34a92b92a1a8204e1d56b6ad05cfd27b/scene/gui/spin_box.cpp#L111-L118

---

What's strange is it works just fine for the FPS SpinBox from the same plugin, which is setuped seemingly the same way:
https://github.com/godotengine/godot/blob/8c729f0f34a92b92a1a8204e1d56b6ad05cfd27b/editor/plugins/sprite_frames_editor_plugin.cpp#L1677-L1685

**Comment by AThousandShips** [comment]
Oh damn my bad, that *is* strange

**Comment by korypostma** [comment]
Yeah I replicated this bug the first time I tried, now hopefully from within the debugger...

**Comment by korypostma** [comment]
I'm able to replicate this in a debug build as long as I attach the debugger after starting, I was not able to replicate it from debugging within VS.  From what I can tell it seems that the `SpinBox` loses focus and the mouse pointer because moving the mouse while holding down the left mouse button no longer modifies the value and it appears like the main viewport took over control / focus but then the input system is left in a limbo state and the only way to get my mouse cursor back was to force the issue within VS's intermediate window.

Do you guys have any ideas on how best to test this to really see what's going on or if the viewport is stealing focus away from the SpinBox?

**Comment by korypostma** [comment]
This appears to be a combination of things.

1. `CanvasItemEditorViewport` is grabbing focus from the `SpinBox` while `MOUSE_CAPTURED` mode is enabled.
2. The `LineEdit` control within the `SpinBox` is the one receiving and responding to messages, so `SpinBox` never receives messages when it loses focus, etc., hence it is unable to call `_release_mouse` when it normally should be able to.

I'm still digging into this, but yeah, this is not that easy of a fix.

**Comment by korypostma** [comment]
Well, I need to do more testing, but if I remove the call to `hide();` in `SpriteFramesEditor::edit(...)` then the control acts like it should.  The call to hide this control was resulting in the `LineEdit` to lose focus, hence allowing the Viewport and others to take focus, and notifications were not always acted upon.

![image](https://github.com/godotengine/godot/assets/5207577/4f48c8c0-aa21-41d8-8565-069211868bb2)

**Comment by korypostma** [comment]
Checking if the SpinBox LineEdit's are in focus, if the mouse is captured, if either SpinBox is in dragging mode all seems to prevent this issue.  I think the only one we really need to check though is if either are in dragging mode (I created this function to access this in SpinBox though).

```
void SpriteFramesEditor::edit(Ref<SpriteFrames> p_frames) {
	_update_stop_icon();

	Control *focus = get_viewport()->gui_get_focus_owner();
	const bool is_focus_on_text = (focus != nullptr && ((focus == frame_duration->get_line_edit()) || (focus == anim_speed->get_line_edit())));
	const bool is_mouse_captured = Input::get_singleton()->get_mouse_mode() == Input::MOUSE_MODE_CAPTURED;
	const bool is_dragging = frame_duration->is_dragging() || anim_speed->is_dragging();
	print_line("is_focus_on_text=", is_focus_on_text, " is_mouse_captured=", is_mouse_captured, " is_dragging=", is_dragging);

	if (!p_frames.is_valid()) {
		if (is_focus_on_text || is_mouse_captured || is_dragging) {
			return;
		}

		frames.unref();
		hide();
		return;
	}
```

**Comment by korypostma** [comment]
I looked into this again and long story short, it is actually only the `is_focus_on_text` variable in the last comment that matters.  The other two checks it will still occur, but if `is_focus_on_text` is true then it will never happen.  I'll modify my source code and provide a pull-request linking this issue and possibly the older one.  I think there is a lot of work to be done on the GUI side of the house though, this could all be handled so much better (most notably the `MOUSE_CAPTURE` stuff).

**Comment by korypostma** [comment]
Another trip through this issue and this is what I have discovered...

Upon creation an `ObjectID` is assigned to the plugin with this value: `SpriteFramesEditorPlugin _instance_id.id: 912798038723`

But in `EditorNode::hide_unused_editors(...)` it is checking `kv.key` is an instance in `ObjectDB` and that value is as follows: `kv.key = 1674684984504`.

Since these keys do not match, it tries to hide the plugin and when the plugin is captured it is then losing the mouse cursor, focus, and other things.

**Comment by ajreckof** [comment]
So after taking a look at it with @korypostma, this is what we have found, there is in fact two bug that results in what we see here : 
- First the issue I created just above which is a SpinBox bug that when it is hidden it does not release the mouse which in turn keep the mouse invisible eternally. This what cause the mouse to stay invisible here.
- Second here more than just two actions happening at the same time because of bad luck it is dragging in itself that triggers the resource to be changed which triggers the update of the whole inspector by AnimatedSprite2d plugin editor. which in turn destroys the EditorPropertyResource editing the SpriteFrames. Therefore the SpriteFrames bottom editor is being hidden.
